### PR TITLE
Add TimestampSelectiveStreamReader

### PR DESCRIPTION
**Pull request by bhhari**
```
Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score    Error  Units
BenchmarkSelectiveStreamReaders.read                  timestamp         true  avgt   60  0.115 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.read                  timestamp        false  avgt   60  0.177 ±  0.002   s/op
BenchmarkSelectiveStreamReaders.readAllNull           timestamp          N/A  avgt   60  0.003 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter        timestamp         true  avgt   60  0.114 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter        timestamp        false  avgt   60  0.198 ±  0.004   s/op

Benchmark                                          Mode  Cnt  Score   Error  Units
BenchmarkBatchStreamReaders.readTimestampNoNull    avgt   60  0.188 ± 0.005   s/op
BenchmarkBatchStreamReaders.readTimestampWithNull  avgt   60  0.166 ± 0.003   s/op
```

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampBatchStreamReader.java [hunk 1]
```diff
@@ -202,40 +203,6 @@ public String toString()
                 .toString();
     }
 
-    // This comes from the Apache Hive ORC code
-    public static long decodeTimestamp(long seconds, long serializedNanos, long baseTimestampInSeconds)
```
Comment (bhhari) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampBatchStreamReader.java]: these methods are moved to a TimestampUtils class as these are used both by the Selective and Batch readers

**Comment by mbasmanova** [comment]
@bhhari Benchmark results suggest that timestamp reading is relatively slow. How do these numbers compare to batch reader?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampBatchStreamReader.java (hunk 1 above)
Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampBatchStreamReader.java]: Could you move this change into a separate commit?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampUtils.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.reader;
+
+public class TimestampUtils
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampUtils.java]: perhaps, make it package private and rename to `ApacheHiveTimestampDecoder`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 3]
```diff
@@ -37,15 +37,17 @@
     private static final Pattern ARRAY_SUBSCRIPT_PATTERN = Pattern.compile("([a-z_+]+)((\\[[0-9]+\\])+)");
 
     private static final String WITH_LINEITEM_EX = "WITH lineitem_ex AS (\n" +
-            "SELECT linenumber, orderkey, partkey, suppkey, \n" +
+            "SELECT linenumber, l.orderkey, partkey, suppkey, \n" +
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: Is there any particular reason to introduce a join? There are 3 date columns in lineitem: shipdate, commitdate, receiptdate. You can convert 2 of these or all 3 to timestamp.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 4]
```diff
@@ -37,15 +37,17 @@
...
             "   CASE WHEN linenumber % 5 = 0 THEN null ELSE shipmode = 'AIR' END AS ship_by_air, \n" +
             "   CASE WHEN linenumber % 7 = 0 THEN null ELSE returnflag = 'R' END AS is_returned, \n" +
-            "   CASE WHEN linenumber % 4 = 0 THEN NULL ELSE CAST(day(shipdate) AS TINYINT) END AS ship_day, " +
-            "   CASE WHEN linenumber % 6 = 0 THEN NULL ELSE CAST(month(shipdate) AS TINYINT) END AS ship_month, " +
-            "   CASE WHEN orderkey % 11 = 0 THEN null ELSE (orderkey, partkey, suppkey) END AS keys, \n" +
-            "   CASE WHEN orderkey % 13 = 0 THEN null ELSE ((orderkey, partkey), (suppkey,), CASE WHEN orderkey % 17 = 0 THEN null ELSE (orderkey, partkey) END) END AS nested_keys, \n" +
-            "   CASE WHEN orderkey % 17 = 0 THEN null ELSE (shipmode = 'AIR', returnflag = 'R') END as flags\n" +
-            "FROM lineitem)\n";
+            "   CASE WHEN linenumber % 4 = 0 THEN null ELSE CAST(day(shipdate) AS TINYINT) END AS ship_day, \n" +
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: this is unrelated change; move it into a separate commit

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/TimestampColumnWriter.java [hunk 5]
```diff
@@ -57,7 +57,7 @@
         implements ColumnWriter
 {
     private static final int INSTANCE_SIZE = ClassLayout.parseClass(TimestampColumnWriter.class).instanceSize();
-    private static final int MILLIS_PER_SECOND = 1000;
+    private static final long MILLIS_PER_SECOND = 1000;
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/writer/TimestampColumnWriter.java]: Why is this change? Could you be compiling with the "wrong" version of JDK?

File: presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkSelectiveStreamReaders.java [hunk 6]
```diff
@@ -289,6 +296,10 @@ private final Object createValue()
                 return new SqlDate(random.nextInt());
             }
 
+            if (getType() == TIMESTAMP) {
+                return new SqlTimestamp(START_TIME.plus(random.nextLong()).getMillis(), UTC_KEY);
```
Comment (mbasmanova) [inline presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkSelectiveStreamReaders.java]: Why not just `return new SqlTimestamp(random.nextLong());`?

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java [hunk 7]
```diff
@@ -727,6 +727,12 @@ else if (type == DATE) {
                         return false;
                     }
                 }
+                else if (type == TIMESTAMP) {
+                    if (!filter.testLong(((SqlTimestamp) value).getMillisUtc())) {
```
Comment (mbasmanova) [inline presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java]: getMillisUtc()  -> getMillis()

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+            boolean outputRequired,
+            LocalMemoryContext systemMemoryContext)
+    {
+        this.streamDescriptor = requireNonNull(streamDescriptor, "streamDescriptor is null");
+        this.filter = filter.orElse(null);
+        this.outputRequired = outputRequired;
+        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+        this.nullsAllowed = this.filter == null || this.filter.testNull();
+        this.baseTimestampInSeconds = new DateTime(
+                2015,
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java]: would you put these arguments all on one line?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+            Optional<TupleDomainFilter> filter,
+            DateTimeZone hiveStorageTimeZone,
+            boolean outputRequired,
+            LocalMemoryContext systemMemoryContext)
+    {
+        this.streamDescriptor = requireNonNull(streamDescriptor, "streamDescriptor is null");
+        this.filter = filter.orElse(null);
+        this.outputRequired = outputRequired;
+        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+        this.nullsAllowed = this.filter == null || this.filter.testNull();
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java]: This logic needs to be updated to support non-deterministic filters. See BooleanSelectiveStreamReader.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+            if (nulls == null || nulls.length < capacity) {
+                nulls = new boolean[capacity];
+            }
+        }
+    }
+
+    private void ensureOutputPositionsCapacity(int capacity)
+    {
+        if (outputPositions == null || outputPositions.length < capacity) {
+            outputPositions = new int[capacity];
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java]: Could you use the new Arrays.ensureCapacity methods here and in ensureValuesCapacity?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+        return newLease(new LongArrayBlock(positionCount, Optional.ofNullable(includeNulls ? nulls : null), values));
+    }
+
+    private BlockLease newLease(Block block)
+    {
+        valuesInUse = true;
+        return ClosingBlockLease.newLease(block, () -> valuesInUse = false);
+    }
+
+    private void compactValues(int[] positions, int positionCount, boolean compactNulls)
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java]: There are multiple readers that store data in array of longs. Would it be make sense to extract this method and logic in getBlock into a utility class to avoid copy-paste?

**Comment by bhhari** [comment]
> @bhhari Benchmark results suggest that timestamp reading is relatively slow. How do these numbers compare to batch reader?

The SelectiveReader is performing better in both the cases
```
Benchmark                                          Mode  Cnt  Score   Error  Units
BenchmarkBatchStreamReaders.readTimestampNoNull    avgt   60  0.174 ± 0.002   s/op
BenchmarkBatchStreamReaders.readTimestampWithNull  avgt   60  0.157 ± 0.001   s/op
```
0.113 vs 0.157
0.169 vs 0.174

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java (hunk 3 above)
Comment (bhhari) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: wanted to compare shipdate vs orderdate to keep it more meaningful

File: presto-orc/src/main/java/com/facebook/presto/orc/writer/TimestampColumnWriter.java (hunk 5 above)
Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/writer/TimestampColumnWriter.java]: it was some refactor i was trying, will revert

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampBatchStreamReader.java (hunk 1 above)
Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampBatchStreamReader.java]: will make a new commit for this

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java (hunk 7 above)
Comment (bhhari) [reply presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java]: looks like this is how its being written, https://github.com/prestodb/presto/blob/6b05323bff91fb089b215737676819817da2befb/presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java#L947
I can make a commit later to remove the usage of the deprecated method. As of now its being used in lot of places which needed to be refactored in a new commit.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java (hunk 11 above)
Comment (bhhari) [reply presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java]: would like to take care of this in a new commit, once the scalars are in I can do a quick refactor there. what do you think?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java (hunk 3 above)
Comment (bhhari) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: Is it ok to keep the join ?

File: presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkSelectiveStreamReaders.java (hunk 6 above)
Comment (bhhari) [reply presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkSelectiveStreamReaders.java]: I need to pass the timeZone as the OrcTester writeLong is using getMillisUtc. I can get rid of the START_TIME though, used it to make it at least more than current time.
https://github.com/prestodb/presto/blob/6b05323bff91fb089b215737676819817da2befb/presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java#L947

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 12]
```diff
@@ -144,6 +148,30 @@ public void testNumeric()
...
+    public void testTimestamps()
+    {
+        // Single timestamp column
+        assertQueryUsingH2Cte("SELECT ship_timestamp FROM lineitem_ex");
+
+        assertQueryUsingH2Cte("SELECT ship_timestamp FROM lineitem_ex WHERE ship_timestamp < TIMESTAMP '1993-01-01 01:00:00'");
+
+        assertQueryUsingH2Cte("SELECT count(*) FROM lineitem_ex WHERE ship_timestamp IS NOT NULL");
+
+        assertQueryUsingH2Cte("SELECT count(*) FROM lineitem_ex WHERE ship_timestamp = (TIMESTAMP '2012-08-08 01:00:00')");
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: parentheses around `TIMESTAMP ...` are not needed

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 13]
```diff
@@ -144,6 +148,30 @@ public void testNumeric()
...
+
+        // Two timestamp columns
+        assertQueryUsingH2Cte("SELECT commit_timestamp, ship_timestamp FROM lineitem_ex");
+
+        assertQueryUsingH2Cte("SELECT commit_timestamp, ship_timestamp FROM lineitem_ex WHERE ship_timestamp > (TIMESTAMP '1993-08-08 01:00:00') AND commit_timestamp < (TIMESTAMP '1993-08-08 01:00:00')");
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: parentheses around `TIMESTAMP ...` are not needed

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 14]
```diff
@@ -144,6 +148,30 @@ public void testNumeric()
...
+
+        assertQueryUsingH2Cte("SELECT commit_timestamp, ship_timestamp FROM lineitem_ex WHERE year(ship_timestamp) - year(commit_timestamp) > 1");
+
+        assertQueryUsingH2Cte("SELECT COUNT(*) FROM lineitem_ex WHERE commit_timestamp IS NULL");
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: this is essentially the same query as in L159; is there any particular reason for duplication?

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java (hunk 7 above)
Comment (mbasmanova) [reply presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java]: @bhhari Sounds good. Thanks for explaining.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java [hunk 15]
```diff
@@ -40,8 +40,10 @@
             "SELECT linenumber, orderkey, partkey, suppkey, \n" +
             "   CASE WHEN linenumber % 5 = 0 THEN null ELSE shipmode = 'AIR' END AS ship_by_air, \n" +
             "   CASE WHEN linenumber % 7 = 0 THEN null ELSE returnflag = 'R' END AS is_returned, \n" +
-            "   CASE WHEN linenumber % 4 = 0 THEN NULL ELSE CAST(day(shipdate) AS TINYINT) END AS ship_day, " +
-            "   CASE WHEN linenumber % 6 = 0 THEN NULL ELSE CAST(month(shipdate) AS TINYINT) END AS ship_month, " +
+            "   CASE WHEN linenumber % 4 = 0 THEN null ELSE CAST(day(shipdate) AS TINYINT) END AS ship_day, " +
```
[the remark this answers is not in the document]
Comment (bhhari) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: Not sure, should be some whitespace

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java (hunk 14 above)
Comment (bhhari) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java]: will remove it
